File: today/binary-tree-zigzag-level-order-traversal.ts
/**
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 */
export { zigzagLevelOrder, TreeNode };

function zigzagLevelOrder(root: TreeNode | null): number[][] {
  if (!root) return [];

  const map = new Map<number, number[]>();

  function dfs(node: TreeNode, level: number) {
    let result = map.get(level);

    if (!result) {
      result = [];
      map.set(level, result);
    }

    result.push(node.val);

    if (node.left) dfs(node.left, level + 1);
    if (node.right) dfs(node.right, level + 1);
  }

  dfs(root, 0);

  const result = Array.from(map.values());

  for (let i = 1; i < result.length; i += 2) {
    result[i].reverse();
  }

  return result;
}

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

File: returned/python/today8/binary_tree_zigzag_level_order_traversal.py
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
from typing import Optional, List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        result = []
        queue = deque([root])
        left_to_right = True
        while queue:
            level_result = []
            for _ in range(len(queue)):
                next = queue.popleft()
                if next.left:
                    queue.append(next.left)
                if next.right:
                    queue.append(next.right)
                level_result.append(next.val)
            if left_to_right:
                result.append(level_result)
            else:
                result.append(level_result[::-1])
            left_to_right = not left_to_right
        return result